test/validation/api/dma/dma.c

```diff
@@ -380,126 +379,132 @@ static int do_transfer(odp_dma_t dma, const odp_dma_transfer_param_t *trs_param,
 	return ret;
 }
 
-static int do_transfer_async(odp_dma_t dma, const odp_dma_transfer_param_t *trs_param,
+static int do_transfer_async(odp_dma_t dma, odp_dma_transfer_param_t *trs_param,
 			     odp_dma_compl_mode_t compl_mode, int multi)
 {
-	odp_dma_compl_param_t compl_param;
+	int num_trs = multi ? multi : 1;
+	odp_dma_compl_param_t compl_param[num_trs];
+	const odp_dma_compl_param_t *compl_ptr[num_trs];
+	const odp_dma_transfer_param_t *trs_ptr[num_trs];
 	odp_event_t ev;
 	odp_dma_compl_t compl;
-	int i, ret, done;
+	int i, j, ret, done;
 	uint32_t user_data = USER_DATA;
 	odp_dma_result_t result;
-	odp_dma_transfer_id_t transfer_id = ODP_DMA_TRANSFER_ID_INVALID;
 	uint64_t wait_ns = 500 * ODP_TIME_MSEC_IN_NS;
 	uint64_t sched_wait = odp_schedule_wait_time(wait_ns);
 	void *user_ptr = &user_data;
 
-	odp_dma_compl_param_init(&compl_param);
-	compl_param.compl_mode = compl_mode;
-
-	if (compl_mode == ODP_DMA_COMPL_EVENT) {
-		compl = odp_dma_compl_alloc(global.compl_pool);
-
-		CU_ASSERT(compl != ODP_DMA_COMPL_INVALID);
-		if (compl == ODP_DMA_COMPL_INVALID)
+	for (i = 0; i < num_trs; i++) {
+		odp_dma_compl_param_init(&compl_param[i]);
+		compl_param[i].compl_mode = compl_mode;
+
+		if (compl_mode == ODP_DMA_COMPL_EVENT) {
+			compl = odp_dma_compl_alloc(global.compl_pool);
+
+			CU_ASSERT(compl != ODP_DMA_COMPL_INVALID);
+			if (compl == ODP_DMA_COMPL_INVALID)
+				return -1;
+
+			compl_param[i].event = odp_dma_compl_to_event(compl);
+			compl_param[i].queue = global.queue;
+		} else if (compl_mode == ODP_DMA_COMPL_POLL) {
+			compl_param[i].transfer_id = odp_dma_transfer_id_alloc(dma);
+
+			CU_ASSERT(compl_param[i].transfer_id != ODP_DMA_TRANSFER_ID_INVALID);
+			if (compl_param[i].transfer_id == ODP_DMA_TRANSFER_ID_INVALID)
+				return -1;
+		} else if (compl_mode != ODP_DMA_COMPL_NONE) {
+			ODPH_ERR("Wrong compl mode: %u\n", compl_mode);
 			return -1;
+		}
 
-		compl_param.event = odp_dma_compl_to_event(compl);
-		compl_param.queue = global.queue;
-	} else if (compl_mode == ODP_DMA_COMPL_POLL) {
-		transfer_id = odp_dma_transfer_id_alloc(dma);
-
-		CU_ASSERT(transfer_id != ODP_DMA_TRANSFER_ID_INVALID);
-		if (transfer_id == ODP_DMA_TRANSFER_ID_INVALID)
-			return -1;
+		compl_param[i].user_ptr = user_ptr;
 
-		compl_param.transfer_id = transfer_id;
-	} else if (compl_mode != ODP_DMA_COMPL_NONE) {
-		ODPH_ERR("Wrong compl mode: %u\n", compl_mode);
-		return -1;
+		if (multi) {
+			trs_ptr[i] = &trs_param[i];
+			compl_ptr[i] = &compl_param[i];
+		}
 	}
 
 	for (i = 0; i < RETRIES; i++) {
-		compl_param.user_ptr = user_ptr;
-
-		if (multi) {
-			const odp_dma_compl_param_t *compl_ptr[1] = {&compl_param};
-			const odp_dma_transfer_param_t *trs_ptr[1] = {trs_param};
-
-			ret = odp_dma_transfer_start_multi(dma, trs_ptr, compl_ptr, 1);
-		} else {
-			ret = odp_dma_transfer_start(dma, trs_param, &compl_param);
-		}
+		if (multi)
+			ret = odp_dma_transfer_start_multi(dma, trs_ptr, compl_ptr, num_trs);
```

Comment:
Depending on the implementation, if `num_trs` is large all transfers may not be started with a single `odp_dma_transfer_start_multi()` call. Perhaps it would be better to check that the return value is  > 0 and repeat call enough times.